README.md
# cloudfront-signer

[![Build Status](https://travis-ci.org/leonelgalan/cloudfront-signer.svg)](https://travis-ci.org/leonelgalan/cloudfront-signer)
[![Code Climate](https://codeclimate.com/github/leonelgalan/cloudfront-signer/badges/gpa.svg)](https://codeclimate.com/github/leonelgalan/cloudfront-signer)
[![Test Coverage](https://codeclimate.com/github/leonelgalan/cloudfront-signer/badges/coverage.svg)](https://codeclimate.com/github/leonelgalan/cloudfront-signer/coverage)
[![Gem Version](https://badge.fury.io/rb/cloudfront-signer.svg)](http://badge.fury.io/rb/cloudfront-signer)
[![Dependency Status](https://gemnasium.com/leonelgalan/cloudfront-signer.svg)](https://gemnasium.com/leonelgalan/cloudfront-signer)

See the [CHANGELOG](https://github.com/leonelgalan/cloudfront-signer/blob/master/CHANGELOG.md)
for details of this release.

See Amazon docs for [Serving Private Content through CloudFront](http://docs.amazonwebservices.com/AmazonCloudFront/latest/DeveloperGuide/index.html?PrivateContent.html)

A fork and rewrite started by [Anthony Bouch](https://github.com/58bits) of
Dylan Vaughn's [aws_cf_signer](https://github.com/dylanvaughn/aws_cf_signer).

This version uses all class methods and a configure method to set options.

Separate helper methods exist for safe signing of urls and stream paths, each of
which has slightly different requirements. For example, urls must not contain
any spaces, whereas a stream path might. Also we might not want to html escape a
url or path if it is being supplied to a JavaScript block or Flash object.

## Installation

This gem has been publised as _cloudfront-signer_. Use `gem install
cloudfront-signer` to install this gem.

The signing class must be configured - supplying the path to a signing key, or
supplying the signing key directly as a string along with the `key_pair_id`.
Create the initializer by running:

```sh
bundle exec rails generate cloudfront:install
```

Customize the resulting *config/initializers/cloudfront\_signer.rb* file.

### Generated *cloudfront\_signer.rb*

```ruby
Aws::CF::Signer.configure do |config|
  config.key_path = '/path/to/keyfile.pem'
  # or config.key = ENV.fetch('PRIVATE_KEY')
  config.key_pair_id  = 'XXYYZZ'
  config.default_expires = 3600
end
```

## Usage

Call the class `sign_url` or `sign_path` method with optional policy settings.

```ruby
Aws::CF::Signer.sign_url 'http://mydomain/path/to/my/content'
```

```ruby
Aws::CF::Signer.sign_path 'path/to/my/content', expires: Time.now + 600
```

Both `sign_url` and `sign_path` have _safe_ versions that HTML encode the result
allowing signed paths or urls to be placed in HTML markup. The 'non'-safe
versions can be used for placing signed urls or paths in JavaScript blocks or
Flash params.

___

Call class method `signed_params` to get raw parameters. These values can be
used to set signing cookies (
[Serving Private Content through CloudFront: Using Signed Cookies](http://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/private-content-signed-cookies.html)
). See [commit message](https://github.com/leonelgalan/cloudfront-signer/commit/fedcc3182e32133e4bd0ad0b79c0106168896c91)
for additional details.

```ruby
Aws::CF::Signer.signed_params 'path/to/my/content'
```

### Custom Policies

See Example Custom Policy 1 at above AWS doc link

```ruby
url = Aws::CF::Signer.sign_url 'http://d604721fxaaqy9.cloudfront.net/training/orientation.avi',
                               expires: 'Sat, 14 Nov 2009 22:20:00 GMT',
                               resource: 'http://d604721fxaaqy9.cloudfront.net/training/*',
                               ip_range: '145.168.143.0/24'
```

With multiple resources

```ruby
url = Aws::CF::Signer.sign_url 'http://d604721fxaaqy9.cloudfront.net/training/orientation.avi',
                               expires: 'Sat, 14 Nov 2009 22:20:00 GMT',
                               resource: [ 'http://d604721fxaaqy9.cloudfront.net/training/*', 'http://d604721fxaaqy9.cloudfront.net/lessons/*' ],
                               ip_range: '145.168.143.0/24'
```

See Example Custom Policy 2 at above AWS doc link

```ruby
Aws::CF::Signer.sign_url 'http://d84l721fxaaqy9.cloudfront.net/downloads/pictures.tgz',
                         starting: 'Thu, 30 Apr 2009 06:43:10 GMT',
                         expires: 'Fri, 16 Oct 2009 06:31:56 GMT',
                         resource: 'http://*',
                         ip_range: '216.98.35.1/32'
```

With multiple resources

```ruby
Aws::CF::Signer.sign_url 'http://d84l721fxaaqy9.cloudfront.net/downloads/pictures.tgz',
                         starting: 'Thu, 30 Apr 2009 06:43:10 GMT',
                         expires: 'Fri, 16 Oct 2009 06:31:56 GMT',
                         resource: [ 'http://d604721fxaaqy9.cloudfront.net/training/*', 'http://d604721fxaaqy9.cloudfront.net/lessons/*' ],
                         ip_range: '216.98.35.1/32'
```

You can also pass in a path to a policy file. This will supersede any other
policy options

```ruby
Aws::CF::Signer.sign_url 'http://d84l721fxaaqy9.cloudfront.net/downloads/pictures.tgz',
                         policy_file: '/path/to/policy/file.txt'
```

## Patches/Pull Requests

* Fork the project.
* Make your feature addition or bug fix.
* Add tests for it.
* Commit
* Send me a pull request. Bonus points for topic branches.

## Attributions

Hat tip to [Anthony Bouch](https://github.com/58bits) for contributing to
Dylan's effort. Only reading both gem's code I was able to figure out the
signing needed for the newly introduced signed cookies.

> Dylan blazed a trail here - however, after several attempts, I was unable to
contact Dylan in order to suggest that we combine our efforts to produce a
single gem - hence the re-write and new gem here. - _Anthony Bouch_

Parts of signing code taken from a question on
[Stack Overflow](http://stackoverflow.com/questions/2632457/create-signed-urls-for-cloudfront-with-ruby)
asked by [Ben Wiseley](http://stackoverflow.com/users/315829/ben-wiseley), and
answered by [Blaz Lipuscek](http://stackoverflow.com/users/267804/blaz-lipuscek)
and [Manual M](http://stackoverflow.com/users/327914/manuel-m).

## License

_cloudfront-signer_ is distributed under the MIT License, portions copyright ©
2015 Dylan Vaughn, STL, Anthony Bouch, Leonel Galán


CHANGELOG.md
# Change Log

## 3.0.1 / 2017-01-20

* Supports signing frozen strings. Bug reported by [@alexandermayr](https://github.com/alexandermayr).

## 3.0.0 / 2015-03-14

* Renames namespace to `Aws`. Matches used by latest [https://github.com/aws/aws-sdk-ruby](https://github.com/aws/aws-sdk-ruby).
  Change proposed by [@tennantje](https://github.com/tennantje)
* Renames `sign` to `build_url` to better communicate method intent.

## 2.2.0 / 2015-04-29

* Accepted merge request from [@leonelgalan](https://github.com/leonelgalan) -
  `sign_params` method returns raw params to be used in urls or cookies.

## 2.1.2 / 2015-04-16

* Accepted merge request from [@tuvistavie](https://github.com/tuvistavie) -
  fixing custom policy bug.

## 2.1.1 / 2013-10-31

* Added changelog file
* Aceppted merge request from [@bullfight](https://github.com/bullfight),
  Refactored configuration to allow for key to be passed in directly.


spec/signer_spec.rb
require 'spec_helper'

RSpec.shared_examples 'is configured' do
  it 'is configured' do
    expect(Aws::CF::Signer.is_configured?).to be true
  end

  it 'sets the private_key' do
    expect(Aws::CF::Signer.send(:private_key)).to(
      be_an_instance_of(OpenSSL::PKey::RSA)
    )
  end
end

FILES_PATH = File.expand_path(File.dirname(__FILE__) + '/files')
KEY_PAIR_ID = 'APKAIKUROOUNR2BAFUUU'.freeze

RSpec.describe Aws::CF::Signer do
  let(:key_path) { FILES_PATH + "/pk-#{KEY_PAIR_ID}.pem" }
  let(:other_key_path) { FILES_PATH + '/private_key.pem' }
  let(:key) { File.readlines(key_path).join '' }

  describe 'Errors' do
    it 'raises ArgumentError when invalid path is passed to key_path' do
      expect do
        Aws::CF::Signer.configure { |config| config.key_path = 'foo/bar' }
      end.to raise_error ArgumentError
    end

    it 'raises OpenSSL::PKey::RSAError when invalid key is passed' do
      expect do
        Aws::CF::Signer.configure { |config| config.key = '' }
      end.to raise_error OpenSSL::PKey::RSAError
    end

    it 'raises ArgumentError when no key is provided through private_key' do
      expect do
        Aws::CF::Signer.configure { |_config| }
      end.to raise_error ArgumentError
    end

    it "raises ArgumentError when no key is provided through key_path doesn't" \
         'allow to guess key_pair_id' do
      expect do
        Aws::CF::Signer.configure { |config| config.key_path = other_key_path }
      end.to raise_error ArgumentError
    end
  end

  describe 'Defaults' do
    it 'expire urls and paths in one hour by default' do
      expect(Aws::CF::Signer.default_expires).to eq 3600
    end

    it 'expires when specified' do
      Aws::CF::Signer.default_expires = 600
      expect(Aws::CF::Signer.default_expires).to eq 600
      Aws::CF::Signer.default_expires = nil
    end
  end

  context 'When configured with key and key_pair_id' do
    before do
      Aws::CF::Signer.configure do |config|
        config.key_pair_id = KEY_PAIR_ID
        config.key = key
      end
    end

    include_examples 'is configured'
  end

  context 'When configured with key_path' do
    before(:each) do
      Aws::CF::Signer.configure { |config| config.key_path = key_path }
    end

    describe 'before default use' do
      include_examples 'is configured'
    end

    describe 'when signing a url' do
      let(:url) { 'https://example.com/someresource?opt1=one&opt2=two' }
      let(:url_with_spaces) { 'http://example.com/sign me' }

      it "doesn't modifies the passed url" do
        url = 'http://example.com/'.freeze
        expect(Aws::CF::Signer.sign_url(url)).not_to match(/\s/)
      end

      it 'removes spaces' do
        expect(Aws::CF::Signer.sign_url(url_with_spaces)).not_to match(/\s/)
      end

      it "doesn't HTML encode the signed url by default" do
        expect(Aws::CF::Signer.sign_url(url)).to match(/\?|=|&/)
      end

      it 'HTML encodes the signed url when using sign_url_safe' do
        expect(Aws::CF::Signer.sign_url_safe(url)).not_to match(/\?|=|&/)
      end
    end

    describe 'when signing a path' do
      it "doesn't remove spaces" do
        path = '/prefix/sign me'
        expect(Aws::CF::Signer.sign_path(path)).to match(/\s/)
      end

      it 'HTML encodes the signed path when using sign_url_safe' do
        path = '/prefix/sign me?'
        expect(Aws::CF::Signer.sign_path_safe(path)).not_to match(/\?|=|&/)
      end
    end

    describe ':expires option' do
      subject(:sign_url) { Aws::CF::Signer.sign_url '', expires: expires }

      { 'Time' => Time.now,
        'String' => '2018-01-01',
        'Fixnum' => 1_514_782_800,
        'NilClass' => nil }.each do |klass, value|
        context "as a #{klass}" do
          let(:expires) { value }
          it "doesn't raise an error" do
            expect { subject }.not_to raise_error
          end
        end
      end

      context 'not as a String, Fixnum or Time' do
        let(:expires) { [[], {}, true, 1.0].sample }
        it 'raises ArgumentError' do
          expect { subject }.to raise_error ArgumentError
        end
      end
    end

    describe 'Custom Policy' do
      it 'builds policy from policy_options' do
        signed_url = Aws::CF::Signer.sign_url(
          'https://d84l721fxaaqy9.cloudfront.net/downloads/pictures.tgz',
          starting: 'Thu, 30 Apr 2009 06:43:10 GMT',
          expires: 'Fri, 16 Oct 2009 06:31:56 GMT',
          resource: 'https://d84l721fxaaqy9.cloudfront.net/downloads/',
          ip_range: '216.98.35.1/32'
        )
        policy_value = get_query_value(signed_url, 'Policy')
        expect(policy_value).not_to be_empty
      end

      it 'builds policy from policy_file' do
        signed_url = Aws::CF::Signer.sign_url(
          'https://d84l721fxaaqy9.cloudfront.net/downloads/pictures.tgz',
          policy_file: FILES_PATH + '/custom_policy.json'
        )
        policy_value = get_query_value(signed_url, 'Policy')
        expect(policy_value).not_to be_empty
      end
    end
  end
end


spec/spec_helper.rb
require 'simplecov'
SimpleCov.start

$LOAD_PATH.unshift(File.join(File.dirname(__FILE__), '..', 'lib'))
$LOAD_PATH.unshift(File.dirname(__FILE__))

require 'rspec'
require 'cloudfront-signer'

def get_query_value(url, key)
  query_string = url.slice((url =~ /\?/) + 1..-1)
  pairs = query_string.split('&')
  pairs.each do |item|
    return item.split('=')[1] if item.start_with?(key)
  end
end


lib/aws-cf-signer.rb
# A re-write of https://github.com/stlondemand/aws_cf_signer
#
require 'openssl'
require 'time'
require 'base64'
require "aws-cf-signer/version"

module AWS
    module CF
        class Signer
            # Public non-inheritable class accessors
            #----------------------------------------------------------------------------
            class << self
                attr_accessor :key_pair_id

                def key_path
                    @key_path
                end

                def key_path=(path)
                    raise ArgumentError.new("The signing key could not be found at #{path}") unless File.exists?(path)
                    @key_path = path
                    @key = OpenSSL::PKey::RSA.new(File.readlines(path).join(""))
                end

                def default_expires
                    @default_expires ||= 3600
                end

                def default_expires=(value)
                    raise ArgumentError.new("The default expires value must be an integer") unless value.is_a?(Integer)
                    @default_expires = value
                end

                private

                # Private key
                #----------------------------------------------------------------------------
                def private_key
                    @key
                end
            end

            #----------------------------------------------------------------------------
            def self.is_configured?
                if self.key_path && self.key_pair_id && private_key
                    return true
                else
                    return false
                end 
            end


            # Configure the signing class. 
            #----------------------------------------------------------------------------
            def self.configure!(key_path, options = {})

                raise ArgumentError.new("You must supply the path to a PEM format private RSA key.") if key_path.nil?
                self.key_path = key_path

                @key_pair_id = options[:key_pair_id] || extract_key_pair_id(key_path)
                unless @key_pair_id
                    raise ArgumentError.new("The Cloudfront signing key id could not be inferred from #{key_path}. Please supply the key pair id as a configuration arguemet.")
                end

                @default_expires = options[:default_expires] ? options[:default_expires] : 3600
            end


            # Signing methods
            #----------------------------------------------------------------------------

            # Sign a url - encoding any spaces in the url before signing. CloudFront 
            # stipulates that signed URLs must not contain spaces (as opposed to stream
            # paths/filenames which CAN contain spaces). 
            #----------------------------------------------------------------------------
            def self.sign_url(subject, policy_options = {}) 
                self.sign(subject, {:remove_spaces => true}, policy_options) 
            end

            # Sign a url (as above) and HTML encode the result.
            #----------------------------------------------------------------------------
            def self.sign_url_safe(subject, policy_options = {}) 
                self.sign(subject, {:remove_spaces => true, :html_escape => true}, policy_options) 
            end

            # Sign a stream path part or filename (spaces are allowed in stream paths 
            # and so are not removed).
            #----------------------------------------------------------------------------
            def self.sign_path(subject, policy_options ={})
                self.sign(subject, {:remove_spaces => false}, policy_options) 
            end

            # Sign a stream path or filename and HTML encode the result.
            #----------------------------------------------------------------------------
            def self.sign_path_safe(subject, policy_options ={})
                self.sign(subject, {:remove_spaces => false, :html_escape => true}, policy_options) 
            end


            # Sign a subject url or stream resource name with optional configuration and 
            # policy options
            #----------------------------------------------------------------------------
            def self.sign(subject, configuration_options = {}, policy_options = {})

                raise "Configure using AWS::CF.Signer.configure! before signing." unless self.is_configured?

                # If the url or stream path already has a query string parameter - append to that.
                separator = subject =~ /\?/ ? '&' : '?'

                if configuration_options[:remove_spaces]
                    subject.gsub!(/\s/, "%20")
                end

                if policy_options[:policy_file]
                    policy = IO.read(policy_options[:policy_file])
                    result = "#{subject}#{separator}Policy=#{encode_policy(policy)}&Signature=#{create_signature(policy)}&Key-Pair-Id=#{@key_pair_id}"
                else
                    if policy_options.keys.size <= 1
                        # Canned Policy - shorter URL
                        expires_at = epoch_time(policy_options[:expires] || Time.now + self.default_expires)
                        policy = %({"Statement":[{"Resource":"#{subject}","Condition":{"DateLessThan":{"AWS:EpochTime":#{expires_at}}}}]})
                        result = "#{subject}#{separator}Expires=#{expires_at}&Signature=#{create_signature(policy)}&Key-Pair-Id=#{@key_pair_id}"
                    else
                        # Custom Policy
                        resource = policy_options[:resource] || subject
                        policy = generate_custom_policy(resource, policy_options)
                        result = "#{subject}#{separator}Policy=#{encode_policy(policy)}&Signature=#{create_signature(policy)}&Key-Pair-Id=#{@key_pair_id}"
                    end
                end

                if configuration_options[:html_escape]
                    return html_encode(result)    
                else
                    return result
                end
            end

            
            # Private helper methods
            #----------------------------------------------------------------------------
            private


            #----------------------------------------------------------------------------
            def self.generate_custom_policy(resource, options)
                conditions = ["\"DateLessThan\":{\"AWS:EpochTime\":#{epoch_time(options[:expires])}}"]
                conditions << "\"DateGreaterThan\":{\"AWS:EpochTime\":#{epoch_time(options[:starting])}}" if options[:starting]
                conditions << "\"IpAddress\":{\"AWS:SourceIp\":\"#{options[:ip_range]}\"" if options[:ip_range]
                %({"Statement":[{"Resource":"#{resource}","Condition":{#{conditions.join(',')}}}}]})
            end

            #----------------------------------------------------------------------------
            def self.epoch_time(timelike)
                case timelike
                when String then Time.parse(timelike).to_i
                when Time   then timelike.to_i
                else raise ArgumentError.new("Invalid argument - String or Time required - #{timelike.class} passed.")
                end
            end

            #----------------------------------------------------------------------------
            def self.encode_policy(policy)
                url_encode(Base64.encode64(policy))
            end

            #----------------------------------------------------------------------------
            def self.create_signature(policy)
                url_encode(Base64.encode64(private_key.sign(OpenSSL::Digest::SHA1.new, (policy))))
            end

            #----------------------------------------------------------------------------
            def self.extract_key_pair_id(key_path)
                File.basename(key_path) =~ /^pk-(.*).pem$/ ? $1 : nil
            end

            #----------------------------------------------------------------------------
            def self.url_encode(s)
                s.gsub('+','-').gsub('=','_').gsub('/','~').gsub(/\n/,'').gsub(' ','')
            end

            #----------------------------------------------------------------------------
            def self.html_encode(s)
                return s.gsub('?', '%3F').gsub('=', '%3D').gsub('&', '%26')
            end
        end
    end
end


lib/cloudfront-signer.rb
# A re-write of https://github.com/stlondemand/aws_cf_signer
#
require 'openssl'
require 'time'
require 'base64'
require 'cloudfront-signer/version'
require 'json'

module Aws
  module CF
    class Signer
      # Public non-inheritable class accessors
      class << self
        # Public: Provides a configuration option to set the key_pair_id if it
        # has not been inferred from the key_path
        #
        # Examples
        #
        #   Aws::CF::Signer.configure do |config|
        #     config.key_pair_id = "XXYYZZ"
        #   end
        #
        # Returns a String value indicating the current setting
        attr_accessor :key_pair_id

        # Public: Provides a configuration option that sets the key_path
        #
        # Examples
        #
        #   Aws::CF::Signer.configure do |config|
        #     config.key_path = "/path/to/your/keyfile.pem"
        #   end
        #
        # Returns nothing.
        def key_path=(path)
          unless File.exist?(path)
            fail ArgumentError,
                 "The signing key could not be found at #{path}"
          end
          @key_path = path
          self.key = File.readlines(path).join('')
        end

        # Public: Provides a configuration option to set the key directly as a
        # string e.g. as an ENV var
        #
        # Examples
        #
        #   Aws::CF::Signer.configure do |config|
        #     config.key = ENV.fetch('KEY')
        #   end
        # Returns nothing.
        def key=(key)
          @key = OpenSSL::PKey::RSA.new(key)
        end

        # Public: Provides an accessor to the key_path
        #
        # Returns a String value indicating the current setting
        attr_reader :key_path

        # Public: Provides a configuration option that sets the default_expires
        # in milliseconds
        #
        # Examples
        #
        #   Aws::CF::Signer.configure do |config|
        #     config.default_expires = 3600
        #   end
        #
        # Returns nothing.
        attr_writer :default_expires

        # Public: Provides an accessor to the default_expires value
        #
        # Returns an Integer value indicating the current setting
        def default_expires
          @default_expires ||= 3600
        end

        private

        # Private: Provides an accessor to the RSA key value
        #
        # Returns an RSA key pair.
        def private_key
          @key
        end
      end

      # Public: Provides a simple way to configure the signing class.
      #
      # Yields self.
      #
      # Examples
      #
      #   Aws::CF::Signer.configure do |config|
      #     config.key_path = "/path/to/yourkeyfile.pem"
      #     config.key_pair_id  = "XXYYZZ"
      #     config.default_expires = 3600
      #   end
      #
      # Returns nothing.
      def self.configure
        yield self if block_given?

        unless key_path || private_key
          fail ArgumentError,
               'You must supply the path to a PEM format RSA key pair.'
        end

        unless @key_pair_id
          @key_pair_id = extract_key_pair_id(key_path)
          fail ArgumentError,
               'The Cloudfront signing key id could not be inferred from ' \
               "#{key_path}. Please supply the key pair id as a " \
               'configuration argument.' unless @key_pair_id
        end
      end

      # Public: Provides a configuration check method which tests to see
      # that the key_path, key_pair_id and private key values have all been set.
      #
      # Returns a Boolean value indicating that settings are present.
      def self.is_configured?
        (key_pair_id.nil? || private_key.nil?) ? false : true
      end

      # Public: Sign a url - encoding any spaces in the url before signing.
      # CloudFront stipulates that signed URLs must not contain spaces (as
      # opposed to stream paths/filenames which CAN contain spaces).
      #
      # Returns a String
      def self.sign_url(subject, policy_options = {})
        build_url subject, { remove_spaces: true }, policy_options
      end

      # Public: Sign a url (as above) and HTML encode the result.
      #
      # Returns a String
      def self.sign_url_safe(subject, policy_options = {})
        build_url subject, { remove_spaces: true, html_escape: true }, policy_options
      end

      # Public: Sign a stream path part or filename (spaces are allowed in
      # stream paths and so are not removed).
      #
      # Returns a String
      def self.sign_path(subject, policy_options = {})
        build_url subject, { remove_spaces: false }, policy_options
      end

      # Public: Sign a stream path or filename and HTML encode the result.
      #
      # Returns a String
      def self.sign_path_safe(subject, policy_options = {})
        build_url subject,
                  { remove_spaces: false, html_escape: true },
                  policy_options
      end

      # Public: Builds a signed url or stream resource name with optional
      # configuration and policy options
      #
      # Returns a String
      def self.build_url(original_subject, configuration_options = {}, policy_options = {})
        subject = original_subject.dup
        # If the url or stream path already has a query string parameter -
        # append to that.
        separator = subject =~ /\?/ ? '&' : '?'

        subject.gsub!(/\s/, '%20') if configuration_options[:remove_spaces]

        result = subject +
                 separator +
                 signed_params(subject, policy_options).collect do |key, value|
                   "#{key}=#{value}"
                 end.join('&')

        if configuration_options[:html_escape]
          return html_encode(result)
        else
          return result
        end
      end

      # Public: Sign a subject url or stream resource name with optional policy
      # options. It returns raw params to be used in urls or cookies
      #
      # Returns a Hash
      def self.signed_params(subject, policy_options = {})
        result = {}

        if policy_options[:policy_file]
          policy = IO.read(policy_options[:policy_file])
          result['Policy'] = encode_policy(policy)
        else
          policy_options[:expires] = epoch_time(policy_options[:expires] ||
                                                Time.now + default_expires)

          if policy_options.keys.size <= 1
            # Canned Policy - shorter URL
            expires_at = policy_options[:expires]
            policy = %{{"Statement":[{"Resource":"#{subject}","Condition":{"DateLessThan":{"AWS:EpochTime":#{expires_at}}}}]}}
            result['Expires'] = expires_at
          else
            # Custom Policy
            resource = policy_options[:resource] || subject
            policy = generate_custom_policy(resource, policy_options)
            result['Policy'] = encode_policy(policy)
          end
        end

        result.merge 'Signature' => create_signature(policy),
                     'Key-Pair-Id' => @key_pair_id
      end

      private

      def self.generate_custom_policy(resource, options)
        conditions = {
          'DateLessThan' => {
            'AWS:EpochTime' => epoch_time(options[:expires])
          }
        }

        conditions['DateGreaterThan'] = {
          'AWS:EpochTime' => epoch_time(options[:starting])
        } if options[:starting]

        conditions['IpAddress'] = {
          'AWS:SourceIp' => options[:ip_range]
        } if options[:ip_range]

        {
          'Statement' => [{
            'Resource' => resource,
            'Condition' => conditions
          }]
        }.to_json
      end

      def self.epoch_time(timelike)
        case timelike
        when String then Time.parse(timelike).to_i
        when Time   then timelike.to_i
        when Fixnum then timelike
        else fail ArgumentError,
                  'Invalid argument - String, Fixnum or Time required - ' \
                  "#{timelike.class} passed."
        end
      end

      def self.encode_policy(policy)
        url_encode Base64.encode64(policy)
      end

      def self.create_signature(policy)
        url_encode Base64.encode64(
          private_key.sign(OpenSSL::Digest::SHA1.new, (policy))
        )
      end

      def self.extract_key_pair_id(key_path)
        File.basename(key_path) =~ /^pk-(.*).pem$/ ? Regexp.last_match[1] : nil
      end

      def self.url_encode(s)
        s.gsub('+', '-').gsub('=', '_').gsub('/', '~').gsub(/\n/, '')
          .gsub(' ', '')
      end

      def self.html_encode(s)
        s.gsub('?', '%3F').gsub('=', '%3D').gsub('&', '%26')
      end
    end
  end
end
